nav2_route/src/breadth_first_search.cpp

```diff
@@ -0,0 +1,176 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_route/breadth_first_search.hpp"
+
+#include <iostream>
+#include <queue>
+#include <unordered_map>
+
+namespace nav2_route
+{
+
+void BreadthFirstSearch::setCostmap(nav2_costmap_2d::Costmap2D * costmap)
+{
+  std::cout << "Set costmap " << std::endl;
+  costmap_ = costmap;
+
+  int x_size = static_cast<int>(costmap_->getSizeInCellsX());
+  int y_size = static_cast<int>(costmap_->getSizeInCellsY());
+
+
+  max_index_ = x_size * y_size;
+  x_size_ = x_size;
+
+  neighbors_grid_offsets = {-1, +1,
+    -x_size, +x_size,
+    -x_size - 1, -x_size + 1,
+    +x_size - 1, +x_size + 1};
+
+  std::cout << "offset size " << neighbors_grid_offsets.size() << std::endl;
+}
+
+BreadthFirstSearch::NodePtr BreadthFirstSearch::addToGraph(const unsigned int index)
+{
+  auto iter = graph_.find(index);
+  if (iter != graph_.end()) {
+    return &(iter->second);
+  }
+
+  return &(graph_.emplace(index, Node{index, false}).first->second);
+}
+
+void BreadthFirstSearch::setStart(unsigned int mx, unsigned int my)
+{
+  start_ = addToGraph(costmap_->getIndex(mx, my));
+}
+
+void BreadthFirstSearch::setGoals(std::vector<unsigned int> mxs, std::vector<unsigned int> mys)
+{
+  goals_.clear();
+  for (unsigned int i = 0; i < (mxs.size() && mys.size()); ++i) {
+    goals_.emplace_back(addToGraph(costmap_->getIndex(mxs[i], mys[i])));
+  }
+}
+
+bool BreadthFirstSearch::search(Coordinates & closest_goal)
+{
+  // clear the graph
+  std::unordered_map<unsigned int, Node> empty_graph;
+  std::swap(graph_, empty_graph);
+
+  std::queue<NodePtr> queue;
+
+  start_->visited = true;
+  queue.push(start_);
+
+  std::cout << "Start index " << start_->index << std::endl;
+  std::cout << "Goal index " << goals_[0]->index << std::endl;
+
+  int iter = 0;
+  while (!queue.empty()) {
+    iter += 1;
+    std::cout << "Iteration " << iter << std::endl;
+    std::cout << "Queue Size: " << queue.size() << std::endl;
+
+    auto & current = queue.front();
+
+    std::cout << "grab front" << std::endl;
+    std::cout << std::endl;
+
+    Coordinates current_coord = getCoords(current->index);
+    std::cout << "CURRENT NODE: " << current_coord.x << " " << current_coord.y << std::endl;
+
+    queue.pop();
+
+    // Check goals
+    for (const auto & goal : goals_) {
+      std::cout << "Goal index " << goal->index << std::endl;
+      std::cout << "current index " << current->index << std::endl;
+      if (current->index == goal->index) {
+        std::cout << "Found goal" << std::endl;
+        closest_goal = getCoords(current->index);
```

Comment:
Perhaps a more natural API would be to return the index of the goal vector who's goal pose is closest. I'm not 100% sure about that, but something to check into on the actual inputs and outputs that would be most natural in the context of the code this is going to integrate into.

Reply:
yeah, 

I think once the search looks good I am going to fix up the api. I am still not sure if we want to check for multiple goals at the same time or just check one.